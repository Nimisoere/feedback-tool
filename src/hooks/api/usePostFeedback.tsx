import React from 'react'

type Props = {}

const usePostFeedback = (props: Props) => {
  return (
    <div>usePostFeedback</div>
  )
}

export default usePostFeedback