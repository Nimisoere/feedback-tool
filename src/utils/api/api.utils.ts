export const fetcher = async (url: string, init: RequestInit) => {
  try {
    const res = await fetch(url, init)
    if (!res.ok) {
      const error: any = new Error('An error occurred')
      error.info = await res.json()
      error.status = res.status
      throw error
    }
    return await res.json()
  } catch (e: any) {
    throw new Error(e?.message)
  }
}