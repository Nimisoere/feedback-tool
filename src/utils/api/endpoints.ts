export enum API_URLS {
  GET_FEEDBACK = "https://63a0db13e3113e5a5c4857fa.mockapi.io/api/v1/feedback"
}