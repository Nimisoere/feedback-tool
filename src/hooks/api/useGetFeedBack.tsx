import useSWR from 'swr'
import { fetcher } from '../../utils/api/api.utils'
import { API_URLS } from '../../utils/api/endpoints'

export type FeedbackResponse = {
  comment: string;
  createdAt: Date
  email: string
  id: string;
  name: string
  rating: number
}

export const useGetFeedBack = () => {
  const { data, error, isLoading, mutate } = useSWR<FeedbackResponse[]>(`${API_URLS.GET_FEEDBACK}?_sort=createdAt&_order=desc`, fetcher)
  return {
    feedbacks: data,
    isLoading,
    error,
    getFeedback: mutate
  }
}

export default useGetFeedBack;