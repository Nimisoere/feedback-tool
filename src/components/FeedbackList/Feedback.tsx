import React from 'react'
import { FeedbackResponse } from '../../hooks/api/useGetFeedBack'
import { range } from '../../utils/common'

type Props = {
  feedback: FeedbackResponse
}

const Feedback = React.memo(function Feedback({ feedback }: Props) {
  const datestring = new Date(feedback.createdAt)?.toLocaleString()
  return (
    <article key={feedback.id} className='flex flex-col mb-4'>
      <div className="font-medium text-lg mb-2 text-black">
        <p>{feedback.name} <span className='text-slate-600 text-sm'>({feedback.email})</span></p>
      </div>
      <div className="flex items-center mb-2">
        {range(1, 5, 1).map((star) => <span className={`text-2xl ${star <= feedback.rating ? "text-yellow-400" : "text-slate-400"}`} key={star}>&#11089;</span>)}
        <p className="ml-2 text-sm font-medium text-gray-900">{feedback.rating} out of 5</p>
      </div>
      <footer className="mb-2 text-sm text-gray-500"><p>Reviewed on <time dateTime={datestring}>{datestring}</time></p></footer>
      <div className="mb-2 font-light text-gray-500">{feedback.comment}</div>
    </article>
  )
})

export default Feedback