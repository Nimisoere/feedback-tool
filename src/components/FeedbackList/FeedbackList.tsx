import Link from 'next/link';
import React from 'react'
import { KeyedMutator } from 'swr';
import useGetFeedBack, { FeedbackResponse } from '../../hooks/api/useGetFeedBack'
import Skeleton from '../common/Skeleton';
import Feedback from './Feedback';

type RenderResultProps = {
  feedbacks: FeedbackResponse[] | undefined;
  isLoading: boolean;
  error: any;
  getFeedback: KeyedMutator<FeedbackResponse[]>
}

const RenderResult: React.FC<RenderResultProps> = ({ isLoading, error, feedbacks, getFeedback }) => {
  if (isLoading) {
    return (
      <Skeleton />
    )
  }
  if (error) {
    return (
      <div role="status" className="w-full">
        <div className='alert-danger'>
          <span className='font-medium'>{error?.message || "Could not fetch feedback list at this time"}</span>
          <button type='button' onClick={() => getFeedback()} className='btn-danger'>
            Try again
          </button>
        </div>
      </div>
    )
  }
  if (feedbacks) {
    if (!feedbacks.length) {
      return (
        <div role="status" className="w-full">
          <div className='alert-warning'>
            <span className='font-medium'>No feedback posted yet</span>
            <Link href={"/"} className='btn-warning'>
              Submit a feedback
            </Link>
          </div>
        </div>
      )
    }
    return (
      <div>
        {
          feedbacks?.map((feedback) => (
            <Feedback key={feedback.id} feedback={feedback} />
          ))
        }
      </div>
    )
  }
  return null;
}

const FeedbackList = () => {
  const { feedbacks, error, isLoading, getFeedback } = useGetFeedBack();

  return (
    <div className='flex flex-col'>
      <div>
        <div className='mb-8'>
          <h2 className='font-medium text-2xl'>Latest Comments</h2>
        </div>
        <RenderResult getFeedback={getFeedback} error={error} isLoading={isLoading} feedbacks={feedbacks} />
      </div>
    </div>
  )
}

export default FeedbackList