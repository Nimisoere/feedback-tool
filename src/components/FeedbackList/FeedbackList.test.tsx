import React from "react";
import "@testing-library/jest-dom";
import { fireEvent, render, screen } from "@testing-library/react";
import FeedbackList from "./FeedbackList";
import { useGetFeedBack } from "../../hooks/api/useGetFeedBack";
import { mockfeedback } from "../../utils/test/mocks";

const useGetFeedBackModule = { useGetFeedBack };

const spy = jest.spyOn(useGetFeedBackModule, "useGetFeedBack");
const myMockFn = jest.fn();

describe("Feedback List - Fetched Results", () => {
  beforeEach(() => {
    spy.mockReturnValue({
      error: null,
      isLoading: false,
      feedbacks: mockfeedback,
      getFeedback: myMockFn
    });
  });
  it("should match snapshot", async () => {
    const { getByText, container } = render(
      <FeedbackList />
    );
    expect(container).toMatchSnapshot();
  });
});


describe("Feedback List - Loading", () => {
  beforeEach(() => {
    spy.mockReturnValue({
      error: null,
      isLoading: true,
      feedbacks: [],
      getFeedback: myMockFn
    });
  });
  it("should match snapshot", async () => {
    const { getByText, container } = render(
      <FeedbackList />
    );
    expect(container).toMatchSnapshot();
  });
});
