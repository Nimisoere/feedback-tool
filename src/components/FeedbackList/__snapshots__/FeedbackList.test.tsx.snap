// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feedback List - Fetched Results should match snapshot 1`] = `
<div>
  <div
    class="flex flex-col"
  >
    <div>
      <div
        class="mb-8"
      >
        <h2
          class="font-medium text-2xl"
        >
          Latest Comments
        </h2>
      </div>
      <div
        class="animate-pulse"
        role="status"
      >
        <div
          class="h-2.5 bg-gray-200 rounded-full max-w-[250px] mb-4"
        />
        <div
          class="h-2 bg-gray-200 rounded-full max-w-[360px] mb-2.5"
        />
        <div
          class="h-2 bg-gray-200 rounded-full mb-2.5"
        />
        <div
          class="h-2 bg-gray-200 rounded-full max-w-[330px] mb-2.5"
        />
        <div
          class="h-2 bg-gray-200 rounded-full max-w-[300px] mb-2.5"
        />
        <div
          class="h-2 bg-gray-200 rounded-full max-w-[360px]"
        />
        <span
          class="sr-only"
        >
          Loading...
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Feedback List - Loading should match snapshot 1`] = `
<div>
  <div
    class="flex flex-col"
  >
    <div>
      <div
        class="mb-8"
      >
        <h2
          class="font-medium text-2xl"
        >
          Latest Comments
        </h2>
      </div>
      <div
        class="animate-pulse"
        role="status"
      >
        <div
          class="h-2.5 bg-gray-200 rounded-full max-w-[250px] mb-4"
        />
        <div
          class="h-2 bg-gray-200 rounded-full max-w-[360px] mb-2.5"
        />
        <div
          class="h-2 bg-gray-200 rounded-full mb-2.5"
        />
        <div
          class="h-2 bg-gray-200 rounded-full max-w-[330px] mb-2.5"
        />
        <div
          class="h-2 bg-gray-200 rounded-full max-w-[300px] mb-2.5"
        />
        <div
          class="h-2 bg-gray-200 rounded-full max-w-[360px]"
        />
        <span
          class="sr-only"
        >
          Loading...
        </span>
      </div>
    </div>
  </div>
</div>
`;
